pkg/querier/frontend/worker.go

```diff
@@ -118,116 +114,90 @@ func (w *worker) watchDNSLoop(servCtx context.Context) error {
 			switch update.Op {
 			case naming.Add:
 				level.Debug(w.log).Log("msg", "adding connection", "addr", update.Addr)
-				ctx, cancel := context.WithCancel(servCtx)
-				cancels[update.Addr] = cancel
-				w.runMany(ctx, update.Addr)
+				client, err := w.connect(servCtx, update.Addr)
+				if err != nil {
+					level.Error(w.log).Log("msg", "error connecting", "addr", update.Addr, "err", err)
+				}
+
+				w.managers[update.Addr] = newFrontendManager(servCtx, w.log, w.server, client, w.cfg.GRPCClientConfig)
 
 			case naming.Delete:
 				level.Debug(w.log).Log("msg", "removing connection", "addr", update.Addr)
-				if cancel, ok := cancels[update.Addr]; ok {
-					cancel()
+				if mgr, ok := w.managers[update.Addr]; ok {
+					mgr.stop()
 				}
 
 			default:
 				return fmt.Errorf("unknown op: %v", update.Op)
 			}
 		}
+
+		w.resetConcurrency()
 	}
 }
 
-// runMany starts N runOne loops for a given address.
-func (w *worker) runMany(ctx context.Context, address string) {
-	client, err := w.connect(address)
-	if err != nil {
-		level.Error(w.log).Log("msg", "error connecting", "addr", address, "err", err)
-		return
+func (w *worker) connect(ctx context.Context, address string) (FrontendClient, error) {
+	opts := []grpc.DialOption{
+		grpc.WithInsecure(),
+		grpc.WithBlock(),
```

Comment:
Oh you snuck a `WithBlock` in there - why?

Reply:
@tomwilkie 
`DialContext()` is non blocking by default.  Adding the `WithBlock()` option makes it behave like `Dial()` in this regard.

Reply:
Do we want it to block?  I think we're assuming the DNS will remove unhealthy clients, which isn't necessarily the case in non-k8s envs.

Reply:
Reread the docs and you're right.